File: src/parser/ast.py
######################################################
# This file is not used at all in the parser itself. #
# It's mostly used for "documentation" purposes      #
######################################################
from enum import Enum, auto
from typing import List

class NodeType(Enum):
    Statement      = auto(),
    NullLiteral    = auto(),
    NumericLiteral = auto(),
    Identifier     = auto(),
    BinaryExpr     = auto(),

    def __repr__(self):
        return f"{self.name}.{self.value}"
    
class Stmt:
    def __init__(self):
        self.kind: NodeType

class Program(Stmt):
    def __init__(self, body):
        self.kind = "Program"
        self.body: List[Stmt] = []

    def __repr__(self):
        return str(self.body)

# Expressions will have a value at runtime unlike the statements
class Expr(Stmt):
    pass

class BinaryExpr(Expr):
    def __init__(self, lhs, rhs, op):
        self.kind: NodeType.BinaryExpr
        self.lhs = lhs
        self.rhs = rhs
        self.op  = op

# Literal/Primary expression types

class Identifier(Expr):
    def __init__(self, symbol):
        self.kind: NodeType.Identifier
        self.symbol = symbol

class NumericLiteral(Expr):
    def __init__(self, value):
        self.kind: NodeType.NumericLiteral
        self.value = value

class NullLiteral(Expr):
    def __init__(self):
        self.kind: NodeType.NullLiteral
        self.value = "null"
        
__all__ = [
    "Stmt",
    "Program",
    "Expr", 
    "BinaryExpr", 
    "Identifier", 
    "NumericLiteral",
    "NullLiteral"
]

if __name__ == "__main__":
    import os
    raise Exception(f"`{os.path.basename(__file__)} can\'t be ran as `{__name__}`")

File: src/parser/ast.js
const { TokenType } = require("./lexer/tokens")

const NodeType = {
    Statement:      "Statement",
    NullLiteral:    "NullLiteral",
    NumericLiteral: "NumericLiteral",
    Identifier:     "Identifier",
    BinaryExpr:     "BinaryExpr"
}


function Stmt() {
    this.kind = NodeType
    return {
        "kind": this.kind
    }
}

function Program() {
    this.kind = "Program"
    this.body = []

    return {
        "kind": this.kind,
        "body": this.body
    }
}


// Expressions will have a value at runtime unlike the statements
function Expr() {
    return {}
}

function BinaryExpr(lhs, rhs, op) {
    this.kind = NodeType.BinaryExpr
    this.lhs = lhs
    this.rhs = rhs
    this.op = op

    return {
        "kind": this.kind,
        "left": this.lhs,
        "right": this.rhs,
        "op": this.op
    }
}

// ---

// Literal/Primary expression types

function Identifier(value) {
    this.kind = NodeType.Identifier
    this.value = value

    return {
        "kind": this.kind,
        "value": this.value
    }
}

function NumericLiteral(value) {
    this.kind = NodeType.NumericLiteral
    this.value = value

    return {
        "kind": this.kind,
        "value": this.value
    }
}

function NullLiteral() {
    this.kind = NodeType.NullLiteral
    this.value = "null"

    return {
        "kind": this.kind,
        "value": this.value
    }
}

module.exports = {
    Stmt,
    Program,
    Expr,
    BinaryExpr,
    Identifier,
    NumericLiteral,
    NullLiteral
}

if (require.main === module) {
    const process = require("process")
    const path = require("path")

    console.error(`ERROR: \`${path.parse(__filename).name}\` file can\'t be ran as main.`)
    process.exit(1)
}